### ISSUE TITLE
Unable to upload files with http form data request

### ISSUE THREAD

**Issue by AhmedMustafa505** [microsoft/react-native-windows#12168]
### Problem Description

I'm implementing a react-native-windows application, and trying to upload file to our server but every time I get Network Error, however the same code works fine on Android.

P.s: I tried to make an http request for a global api, and I succeeded. so the issue is relying behind the form-data technique or RNW configurations I'm missing.

Also noticed that inside the axios error response that Content-Type is set to false.

### Steps To Reproduce

1. Picked a file using react-native-document-picker.
2. Tried to wrap the file data in a formdata and add it as a body of http request.

### Expected Results

Response of successful request from the server, but got a Network Error.

### CLI version

11.3.6

### Environment

```markdown
System:
  OS: Windows 10 10.0.19045
  CPU: (8) x64 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
  Memory: 6.31 GB / 15.76 GB
Binaries:
  Node:
    version: 18.16.0
    path: C:\Program Files\nodejs\node.EXE
  Yarn:
    version: 1.22.19
    path: C:\Program Files (x86)\Yarn\bin\yarn.CMD
  npm:
    version: 9.7.1
    path: C:\Program Files\nodejs\npm.CMD
  Watchman: Not Found
SDKs:
  Android SDK: Not Found
  Windows SDK:
    AllowDevelopmentWithoutDevLicense: Enabled
    AllowAllTrustedApps: Enabled
    Versions:
      - 10.0.18362.0
      - 10.0.19041.0
      - 10.0.20348.0
      - 10.0.22000.0
      - 10.0.22621.0
IDEs:
  Android Studio: AI-222.4459.24.2221.10121639
  Visual Studio:
    - 17.6.33723.286 (Visual Studio Community 2022)
    - 17.7.33711.374 (Visual Studio Community 2022)
Languages:
  Java: 11.0.2
  Ruby: Not Found
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.72.4
    wanted: 0.72.4
  react-native-windows:
    installed: 0.72.9
    wanted: 0.72.9
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: Not found
  newArchEnabled: Not found


```

### Target Platform Version

10.0.19041

### Target Device(s)

Desktop

### Visual Studio Version

Visual Studio 2022

### Build Configuration

Debug

### Snack, code example, screenshot, or link to a repository

![tempsnip](https://github.com/microsoft/react-native-windows/assets/123844507/3888821e-e9a1-4cbe-a0ed-edd037e76b91)
![Error](https://github.com/microsoft/react-native-windows/assets/123844507/fd35ed66-ad95-4b0f-bf73-400ed728a13c)

**Comment by chrisglein** [comment]
> P.s: I tried to make an http request for a global api, and I succeeded. so the issue is relying behind the form-data technique or RNW configurations I'm missing.

Can you clarify what you tried here? I'm curious if this works for you in a vanilla React web app, for example? Or a non-Windows React Native app? I'm curious what indicated that this is an RNW issue.

Have you looked at all in [react-native-document-picker](https://github.com/rnmods/react-native-document-picker) itself and if it has any known issues here in the Windows implementation?

**Comment by AhmedMustafa505** [comment]
> > P.s: I tried to make an http request for a global api, and I succeeded. so the issue is relying behind the form-data technique or RNW configurations I'm missing.
> 
> Can you clarify what you tried here? I'm curious if this works for you in a vanilla React web app, for example? Or a non-Windows React Native app? I'm curious what indicated that this is an RNW issue.
> 
> Have you looked at all in [react-native-document-picker](https://github.com/rnmods/react-native-document-picker) itself and if it has any known issues here in the Windows implementation?

I tried the same api request in an android react native app with the same code structure and the request was successful.

And also tried to make another request in RNW to a global api to determine whether it’s network problem or not and I succeeded, that’s what indicated the issue maybe related to RNW and multipart post requests

**Comment by JunielKatarn** [comment]
Looking into this.

Will attempt to reproduce locally.

**Comment by JunielKatarn** [comment]
> > > P.s: I tried to make an http request for a global api, and I succeeded. so the issue is relying behind the form-data technique or RNW configurations I'm missing.
> > 
> > 
> > Can you clarify what you tried here? I'm curious if this works for you in a vanilla React web app, for example? Or a non-Windows React Native app? I'm curious what indicated that this is an RNW issue.
> > Have you looked at all in [react-native-document-picker](https://github.com/rnmods/react-native-document-picker) itself and if it has any known issues here in the Windows implementation?
> 
> I tried the same api request in an android react native app with the same code structure and the request was successful.
> 
> And also tried to make another request in RNW to a global api to determine whether it’s network problem or not and I succeeded, that’s what indicated the issue maybe related to RNW and multipart post requests

Can you please clarify what you mean by "global api"?
It would be useful to know what kind of endpoint was used to reproduce this issue.

**Comment by AhmedMustafa505** [comment]
```
const getNews = async => {
    const apiKey = '1fec1ec7232c47cc8d432720df947253';
    const apiUrl = 'http://newsapi.org/v2/everything';
  const queryParameters = {
      q: 'tesla',
      from: '2023-09-5',
      sortBy: 'publishedAt',
      apiKey: apiKey,
    };
    axios
      .get(apiUrl, {params: queryParameters})
      .then(response => {
        // Handle the API response data here
        console.log('API Response:', response.data);
      })
      .catch(error => {
        // Handle any errors that occurred during the request
        console.error('API Request Error:', error);
      });
  };
```

  

```
 const uploadFile = async () => {
    const pickerResult = await DocumentPicker.pickSingle({
      presentationStyle: 'fullScreen',
      copyTo: 'documentDirectory',
      type: [DocumentPicker.types.allFiles],
      mode: 'import',
    });
    console.log(pickerResult);

    const response = uploadImage(pickerResult);
  };
  const uploadImage = async file => {
    const formData = new FormData();
    formData.append('file', {
      uri: file.fileCopyUri,
      type: file.type,
      name: file.name,
      fileName: file.name,
      size: file.size,
    });

    console.log(formData);
    try {
      const response = await axios.post(
        'http://XXX.XXX.XX.XXX:XXX/upload',
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        },
      );

      console.log('Image upload success:', response.data);
    } catch (error) {
      console.error('Image upload error:', error);
    }
  };

```

by "global api" I mean I used getNews global api in the first function to check whether it's network related issue in RNW or not, but it worked fine in both platforms (Android & RNW).

for the second function that uploads files, it's only working on Android.

**Comment by AhmedMustafa505** [comment]
I also tried to upload with a different technique using XMLHttpRequest as follows:

```
 const uploadImage = async file => {
    const xhr = new XMLHttpRequest();
    xhr.open('POST', 'http://***.***.**.***:****/upload', true); 
  
    xhr.upload.onprogress = event => {
      console.log('heree');

      if (event.lengthComputable) {
        const percentCompleted = Math.round((event.loaded / event.total) * 100);
        console.log(`Upload progress: ${percentCompleted}%`);
      }
    };


    xhr.setRequestHeader('Content-Type', 'multipart/form-data');
    const formData = new FormData();
    formData.append('file', {
      uri: file.uri,
      type: file.type,
      name: file.name,
    });
    console.log(formData); 
    try {
      xhr.send(formData);

      xhr.onload = () => {
        console.log('hereee');

        if (xhr.status === 200) {
          console.log('Image upload success:', xhr.responseText);
          // Handle success
        } else {
          console.error('Image upload failed:', xhr.statusText);
        }
      };
    } catch (error) {
      console.error('Image upload error:', error);
    }
  };
```

Got the same result as it succeeded in android but not in RNW.

**Comment by WilliamWelsh** [comment]
I have the same exact issue, can't post files with Axios. Same exact code runs fine on both iOS and Android, but Windows throws a network error

**Comment by gtRfnkN** [comment]
The same issue happens with graphql using the apollo upload client (https://github.com/jaydenseric/apollo-upload-client). Uploading a file (in this case wrapped in `ReactNativeFile`) throws a network error. The same code (server and client) works fine when using a web client or our old windows app running RNW 0.63. The original examples of that library can be found here: https://github.com/jaydenseric/apollo-upload-examples/tree/master/api


I've also created a repository to reproduce this issue: https://github.com/gtRfnkN/rnw-apollo
To try it out, simply install and run the local server located in `/apollo-server`. Then install and run the windows app from root. Once the app has started, click the "load apollo component" button at the top, click "pick and upload file", and choose any file. This should upload the file and show the response from the server. Instead, a network error is thrown.

Caution: this repository was first created for https://github.com/microsoft/react-native-windows/issues/12187. Therefore, it includes another issue which can result in `Error: Invalid response for blob` and `Error: ENOENT: no such file or directory, open '\\?\E:\git\rnw-apollo\code'`. To ignore this, you can simply enable remote js debugging (which will also console.log the error)

![image](https://github.com/microsoft/react-native-windows/assets/2790592/14ff34d1-5022-42e8-b5e9-29559edc2f18)

**Comment by WilliamWelsh** [comment]
I can't find it now, but I remember seeing a photo upload example and it used a native module instead of just axios or fetch or something else. I have a feeling I'm gonna have to add a native module just for Windows just for uploading files.

**Comment by AhmedMustafa505** [comment]
> I can't find it now, but I remember seeing a photo upload example and it used a native module instead of just axios or fetch or something else. I have a feeling I'm gonna have to add a native module just for Windows just for uploading files.

Unfortunately I haven't added native modules myself before, if you managed to add native modules and will do this for the file upload can you share the solution and the steps ?

**Comment by gtRfnkN** [comment]
I've updated [my reproduction repository](https://github.com/gtRfnkN/rnw-apollo) to include a simple expressjs+multer server to handle file uploads. It also delivers a very basic html page to upload a file. Simply run `yarn simple-server` from the root folder and then open `http://localhost:8080/`. Once a file is uploaded, it will be visible inside `/simple-server/uploads` and the server will log the file count via the cli.

In the windows app, there is a new button to select and upload a file to that local server via `fetch` and `FormData`. This fails with the already mentioned `Network request failed`. The server doesn't receive any call and doesn't log any information. The app also has a "upload blob" button to simply upload a string via multipart formdata to make sure it's not a limitation of the file picker or some permission issues. This button throws the same error.

I also tried to debug the native code inside `WinRTHttpResource.cpp` but I couldn't get the symbols to work as I'm not very familiar with Visual Studio nor C# development. But if there is anything else I could provide or look into, please let me know.

**Comment by AhmedMustafa505** [comment]
It's the same approach I took and still didn't result in any successful connection to the server.

**Comment by WilliamWelsh** [comment]
I was able to upload files by writing a native C# module where I pass the file path to it and it uploads (HttpClient), so I'm not able to do any uploading via axios or anything in JS. In my app now, I have something like this:

```ts
if (Platform.OS == "windows") {
      WindowsCameraController.Upload(
        yourFilePath
      );
else
// axios stuff
```

There's some more details [here](https://github.com/WilliamWelsh/WilliamWelsh/issues/1)

**Comment by gtRfnkN** [comment]
Nice work, that sounds like a good workaround.

Is there some insight on when this issue might get fixed inside react-native-windows itself? Passing requests, responses, loading state, etc through a native module compared to a simple fetch is quite some hassle. But I suspect there is no easy way to integrate these C# changes into react-native-windows because that is written in C++?

**Comment by gtRfnkN** [comment]
I tested this issue with different RNW releases and was able to narrow it down.

First of all, this  is not (only) about uploading files. Simply using multipart/form-data will cause this issue. You can reproduce this with the example from the RNW repository: [packages/@react-native-windows/tester/src/js/examples/HTTP/HTTPExampleMultiPartFormData.js](https://github.com/microsoft/react-native-windows/blob/main/packages/%40react-native-windows/tester/src/js/examples/HTTP/HTTPExampleMultiPartFormData.js)

Based on that example, I've created a simple RNW app using fetch with multipart/form-data with and without uploading a file. Here are the results based on the RNW version:
0.70 -> works up until 0.71.24*
0.71 -> works for 0.71.0, 0.71.21, 0.71.24 | **fails for 0.71.25**, 0.71.26, 0.71.31, 0.71.43*
0.72 -> works for 0.72.1 | **fails for 0.72.2**, 0.72.9, 0.72.17*
(* = latest version at the time of writing)

Looking at the PRs for the versions which broke multipart/form-data (0.71.25 & 0.72.2), they both include the same change to TurboModules for networking: https://github.com/microsoft/react-native-windows/pull/11867 (this change PR was also opened for 0.70 but never merged)

Unfortunately, I don't have enough knowledge of C++ to pintpoint the exact issue, but I'm guessing that's where it came from. But rolling the app back to 0.72.1 (or 0.71.24) resolves the issue for now.

**Comment by AhmedMustafa505** [comment]
> Nice work, that sounds like a good workaround.
> 
> Is there some insight on when this issue might get fixed inside react-native-windows itself? Passing requests, responses, loading state, etc through a native module compared to a simple fetch is quite some hassle. But I suspect there is no easy way to integrate these C# changes into react-native-windows because that is written in C++?

My workaround for this was to initiate the project with C# instead of C++ , use this command :
`react-native-windows-init --language cs`

**Comment by tero-paananen** [comment]
Can this issue causes this upload error https://github.com/microsoft/react-native-windows/issues/12392

**Comment by gtRfnkN** [comment]
Yes, that is the root cause of this upload issues.

@AhmedMustafa505 & @WilliamWelsh: there is an explanation of the issue and a patch file to resolve this issue without a custom native module in https://github.com/microsoft/react-native-windows/issues/12392

**Comment by JunielKatarn** [comment]
Solved by #12506.
Thanks to @tero-paananen for the investigation and fix proposal.
